pep-0592.rst

```diff
@@ -0,0 +1,295 @@
+PEP: 592
+Title: External annotations in the typing module
+Author: Till Varoquaux <till@fb.com>, Konstantin Kashin <kkashin@fb.com>
+Sponsor: Ivan Levkivskyi <levkivskyi@gmail.com>
+Discussions-To: typing-sig@python.org
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 26-April-2019
+Python-Version:
+Post-History:
+
+Abstract
+--------
+
+This PEP introduces a mechanism to extend the type annotations from PEP
+484 with arbitrary metadata.
+
+Motivation
+----------
+
+PEP 484 provides a standard semantic for the annotations introduced in
+PEP 3107. PEP 484 is prescriptive but it is the de-facto standard
+for most of the consumers of annotations; in many statically checked
+code bases, where type annotations are widely used, they have
+effectively crowded out any other form of annotation. Some of the use
+cases for annotations described in PEP 3107 (database mapping,
+foreign languages bridge) are not currently realistic given the
+prevalence of type annotations. Furthermore the standardisation of type
+annotations rules out advanced features only supported by specific type
+checkers.
+
+Rationale
+---------
+
+We propose adding an ``Annotated`` type to the typing module to decorate
+existing types with context-specific metadata. Specifically, a type
+``T`` can be annotated with metadata ``x`` via the typehint
+``Annotated[T, x]``. This metadata can be used for either static
+analysis or at runtime. If a library (or tool) encounters a typehint
+``Annotated[T, x]`` and has no special logic for metadata ``x``, it
+should ignore it and simply treat the type as ``T``. Unlike the
+``no_type_check`` functionality that current exists in the ``typing``
+module which completely disables typechecking annotations on a function
+or a class, the ``Annotated`` type allows for both static typechecking
+of ``T`` (e.g., via mypy [mypy]_ or Pyre [pyre]_, which can safely ignore ``x``)
+together with runtime access to ``x`` within a specific application. We
+believe that the introduction of this type would address a diverse set
+of use cases of interest to the broader Python community.
+
+This was originally brought up as issue 600 [issue-600]_ in the typing github
+and then discussed in Python ideas [python-ideas]_.
+
+Motivating examples
+-------------------
+
+Combining runtime and static uses of annotations
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+There's an emerging trend of libraries leveraging the typing annotations at
+runtime (e.g.: dataclasses); having the ability to extend the typing annotations
+with external data would be a great boon for those libraries.
+
+Example::
```

Comment:
Make it clear that this example is about hypothetic functionality in `cstruct` module.